python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py

```diff
@@ -0,0 +1,331 @@
+
+        # Compute the neighbors lists requested by the model using ASE NL
+        for options in self._model.requested_neighbors_lists():
+            neighbors = _compute_ase_neighbors(atoms, options)
+            register_autograd_neighbors(
+                system,
+                neighbors,
+                check_consistency=self.parameters["check_consistency"],
+            )
+            system.add_neighbors_list(options, neighbors)
+
+        return self._model(
+            [system],
+            run_options,
+            check_consistency=self.parameters["check_consistency"],
+        )
+
+    def calculate(
+        self,
+        atoms: ase.Atoms,
+        properties: List[str],
+        system_changes: List[str],
+    ) -> Dict[str, np.ndarray]:
+        """
+        Compute some ``properties`` with this calculator, and return them in the format
+        expected by ASE.
+
+        This is not intended to be called directly by users, but to be an implementation
+        detail of ``atoms.get_energy()`` and related functions. See
+        :py:meth:`ase.calculators.calculator.Calculator.calculate` for more information.
+        """
+        super().calculate(
+            atoms=atoms,
+            properties=properties,
+            system_changes=system_changes,
+        )
+
+        outputs = _ase_properties_to_metatensor_outputs(properties)
+        species, positions, cell = _ase_to_torch_data(atoms)
+
+        do_backward = False
+        if "forces" in properties:
+            do_backward = True
+            positions.requires_grad_(True)
+
+        if "stress" in properties:
+            do_backward = True
+
+            scaling = torch.eye(3, requires_grad=True, dtype=cell.dtype)
+
+            positions = positions @ scaling
+            positions.retain_grad()
+
+            cell = cell @ scaling
+
+        if "stresses" in properties:
+            raise NotImplementedError("'stresses' are not implemented yet")
+
+        # convert from ase.Atoms to metatensor.torch.atomistic.System
+        system = System(species, positions, cell)
+        for options in self._model.requested_neighbors_lists():
+            neighbors = _compute_ase_neighbors(atoms, options)
+            register_autograd_neighbors(
+                system,
+                neighbors,
+                check_consistency=self.parameters["check_consistency"],
+            )
+            system.add_neighbors_list(options, neighbors)
+
+        run_options = ModelEvaluationOptions(
+            length_unit="angstrom",
+            outputs=outputs,
+            selected_atoms=None,
+        )
+
+        outputs = self._model(
+            [system],
+            run_options,
+            check_consistency=self.parameters["check_consistency"],
+        )
+        energy = outputs["energy"]
+
+        if run_options.outputs["energy"].per_atom:
+            assert len(energy) == 1
+            assert energy.sample_names == ["structure", "atom"]
```

Comment:
In our models, it would be easier to have the samples be `["structure", "center"]`. This would be cleaner on the model side, so we don't have to re-create a TensorMap. It also fits multi-center properties perhaps, where the samples might be `["structure", "center_1", "center_2"]`...